Graft uncompressed files via gateway

mcb30 opened the pull request:
Add the ability to flag grafted files as uncompressed, and tests to verify that uncompressed grafted files can be published via a gateway.

mcb30 committed:
Always log decompression errors in downloaded files

Decompression errors should not occur in normal operation: they
represent either data corruption or a misconfiguration.  In either
case, it is useful to have this information immediately visible even
at a non-debug log level.

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

mcb30 committed:
Include explicit compression mechanism within graft files

Files are marked as compressed or uncompressed at the level of
individual catalog entries, and the "cvmfs_swissknife graft" tool is
already capable of creating graft marker files for either compressed
or uncompressed versions of the file.  However, the information about
the compression mechanism is not stored within the graft marker file,
which makes it difficult to manage a repository containing a mixture
of compressed and uncompressed files.

Add a "compression=..." line to the graft marker files and ensure that
the compression method is propagated into the newly created catalog
entry.

For backwards compatibility, treat an absent "compression=..." line as
meaning that the compression mechanism should be inferred from the
sync mediator in use.

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

jblomer commented:
Is this meant to support grafting compressed and uncompressed files in the same transaction? To graft all files uncompressed, it should be enough to set `CVMFS_COMPRESSION_ALGORITHM=none` on the publisher node.

mcb30 commented:
> Is this meant to support grafting compressed and uncompressed files in the same transaction? To graft all files uncompressed, it should be enough to set `CVMFS_COMPRESSION_ALGORITHM=none` on the publisher node.

It does support grafting a mixture of compressed and uncompressed files in the same transaction, but my primary motivation was to make the process of using uncompressed files less confusing by recording the compression information within the graft metadata so that the publisher does the right thing without requiring the user to know about setting `CVMFS_COMPRESSION_ALGORITHM`.  (I personally messed up several times trying to get grafting of uncompressed files to work, and I suspect I'm not alone in this.)

jblomer commented:
Thanks for the clarification! That sounds reasonable and a useful improvement.

jblomer commented:
@cernvm-bot cloudtest

mcb30 commented:
> https://lcgapp-services.cern.ch/cvmfs-jenkins/job/CvmfsCloudTesting/1122/

Looks as though `openssl zlib` doesn't exist as a command on some of the platforms - is there a standard way for CVMFS test scripts to construct compressed versions of files?

Also looks as though the S3 tests are failing because `/srv/cvmfs/<repo>` doesn't exist on those platforms, which is reasonable enough.  Is there a standard way for CVMFS test scripts to inject content-addressed files directly into whatever functions as the storage backend?

Thanks

DrDaveD commented:
> (I personally messed up several times trying to get grafting of uncompressed files to work, and I suspect I'm not alone in this.)

On the contrary you probably are alone on this, since grafting is not a very heavily used feature!  I think you are only the second user, the first being the authors @djw8605 and @bbockelm for osgstorage.org repositories.

jblomer commented:
For the compression: `cat foo | cvmfs_publish zpipe > foo.compressed` and with `cvmfs_publish zpipe -d` for decompression.

For planting a file in the backend, there is `upload_into_backend` in `test_functions`

mcb30 commented:
> For the compression: `cat foo | cvmfs_publish zpipe > foo.compressed` and with `cvmfs_publish zpipe -d` for decompression.
> 
> For planting a file in the backend, there is `upload_into_backend` in `test_functions`

Wonderful; thank you!  Am updating now...

mcb30 commented:
I'm hitting an error that I don't understand:
```
+ cvmfs_swissknife upload -i file1 -o data/bf/c7a4e4317c6601b632801b210d0359a6315a11 -r gw,/srv/cvmfs/test.repo.org/txn,http://localhost:4929/api/v1 -a sha1
Failed to configure gateway uploader. Missing session token file.
```
At the point that the error occurs, the file `/var/spool/cvmfs/test.repo.org/session_token` does exist, and a `cvmfs_server publish test.repo.org` at this point would succeed.

I can't find the portion of code that is supposed to construct the path to the `session_token` file.  The string `"session_token"` appears only twice across the whole C++ codebase, both of which are using it as a key name for `JsonDocument::SearchInObject()`.  Where within the code does anything access this file?

Thanks

jblomer commented:
That's right, the `cvmfs_swissknife upload` command only works with the local and the S3 backend.  It would be better to bail out with an error message when this is called with a gateway uploader.

In the integration tests, the gateway runs on the same machine that the publisher, so we should pick the backend configuration of the gateway.  You can get this original backend from `TEST_CVMFS_RECEIVER_UPSTREAM_STORAGE` in /etc/cvmfs/repositories.d/test.repo.org/server.conf.  It might be best to adjust `upload_into_backend` accordingly.

mcb30 committed:
Allow upload_into_backend() to upload files during a gateway test

The "cvmfs_swissknife upload" command cannot be used to upload files
via a gateway.  Since integration tests always run with both the
gateway and publisher on the same machine, adapt the logic in
upload_into_backend() to perform the upload directly to the backend
storage even when a gateway is in use.

Signed-off-by: Michael Brown <mbrown@fensystems.co.uk>

mcb30 commented:
Thanks.  I've updated to use `cvmfs_publish zpipe` and `upload_into_backend` as appropriate, with a separate commit to update `upload_into_backend` (and related functions) to handle the case of a gateway test.

mcb30 commented:
@cernvm-bot cloudtest

mcb30 commented:
> @cernvm-bot cloudtest

Wasn't sure if that was going to work, and it looks as though @cernvm-bot indeed ignores requests from non-admin users.

@jblomer Could you possibly kick off cloudtests for this?  Thank you!

jblomer commented:
@cernvm-bot cloudtest

mcb30 commented:
> cloudtests finished: FAILURE
> https://lcgapp-services.cern.ch/cvmfs-jenkins/job/CvmfsCloudTesting/1124/

Looks like just a single transient failure in test 633 on Ubuntu 16.04 S3.  AFAICT this test should not be affected by these changes, and the same test passed on other S3 platforms and other Ubuntu 16.04 platforms in this test run.

jblomer commented:
@cernvm-bot unittest

jblomer commented:
@cernvm-bot cpplint

jblomer commented on test/src/811-repository_gateway_graft/main:
```diff
@@ -0,0 +1,81 @@
+  verify_compression ${spooldir}/rdonly/file2 zlib || return $?
+  verify_sha1sum ${repodir}/file1 ${hash1} || return $?
+  verify_sha1sum ${repodir}/file2 ${hash2} || return $?
+}
```
It's not really necessary but in line with the other tests it would be nice to have an explicit `return 0` at the end.

mcb30 replied:
Will do

jblomer commented on cvmfs/sync_union.cc:
```diff
@@ -36,7 +36,9 @@ SharedPtr<SyncItem> SyncUnion::CreateSyncItem(
     entry->SetExternalData(mediator_->IsExternalData());
-    entry->SetCompressionAlgorithm(mediator_->GetCompressionAlgorithm());
+    if (!entry->IsValidGraft()) {
+      entry->SetCompressionAlgorithm(mediator_->GetCompressionAlgorithm());
```
This might affect users who create graft files but do not use the `swissknife_graft` utility.  I'd prefer to make the `compression` key optionally so that graft files without the key would use `mediator_->GetCompressionAlgorithm()`.  Let's add a test for this case.

mcb30 replied:
Good point - I'll add this now.  Thank you!

mcb30 replied:
Added along with a test case

jblomer reviewed (changes requested):
Many thanks! This looks in principle good to me, I have just one concern regarding backwards compatibility of the graft file format.

mcb30 commented:
@jblomer I've updated as per your comments, and rebased onto current `devel`.  Thanks for the review!

jblomer commented:
@cernvm-bot cloudtest

jblomer commented:
@cernvm-bot unittest

jblomer commented:
@cernvm-bot cpplint
